### ISSUE TITLE
Define the placement of data by SQL statements

### ISSUE THREAD

**Issue by scsldb** [pingcap/tidb#18030]
### Description

TiDB supports placement rules, which can define the placement of data in a more flexible and more granular way. But it only provides configuration files to define them, and it’s complicated.

This issue tracks our progress in adding support via SQL.

* [Weekly Report](https://docs.google.com/document/d/1_A-bU2JoXZ0Fllhh0xJUdEtDXJQmz52toeNItW42fF0/edit#heading=h.4djbdyo32xia)
* [RFC document](https://github.com/pingcap/tidb/blob/master/docs/design/2020-06-24-placement-rules-in-sql.md)

The use cases for Placement Rules generally fit into two broad categories:
1. Optimization (local stale reads) and High Availability
2. Compliance / Geographical Restrictions

We are designing the syntax with both of these features in mind, but anticipate that initially use-case (1) will be better handled.

### Named Placement Policy

(Previously: https://github.com/pingcap/tidb/issues/26580 )

* [x] Parser: {CREATE|ALTER} PLACEMENT POLICY placement_options https://github.com/pingcap/parser/pull/1313
* [x] Parser: SHOW CREATE PLACEMENT POLICY x, DROP PLACEMENT POLICY [IF EXISTS] x; https://github.com/pingcap/parser/pull/1295 https://github.com/pingcap/parser/pull/1298
* [x] Design table in mysql.* for persisting placement policies. https://github.com/pingcap/tidb/pull/27251
* [x] Ensure {CREATE|ALTER|DROP} PLACEMENT POLICY requires the PLACEMENT_ADMIN privilege https://github.com/pingcap/tidb/issues/27603
* [x] #27714
* [x] support the inheritance of DB/Table/Partition placement policy https://github.com/pingcap/tidb/pull/28365
* [x] postpone the clean logic of the placement policy to gc phase https://github.com/pingcap/tidb/pull/28687  

### Support SHOW PLACEMENT

(Previously: https://github.com/pingcap/tidb/issues/26582 )

* [x] Parser: `SHOW PLACEMENT LIKE 'x';
* [x] Parser: Add `SHOW PLACEMENT FOR [{DATABASE | SCHEMA} schema_name] [TABLE table_name] [PARTITION partition_name]`
* [x] `SHOW PLACEMENT LIKE` should return a list of `PLACEMENT POLICIES` that have been loaded in the TiDB server.
  - #27531
* [x] `SHOW PLACEMENT FOR` https://github.com/pingcap/tidb/issues/27976
  - #28266
  - #28366
  - #28830 
* [x] Privileges for `SHOW PLACEMENT` and `SHOW PLACEMENT FOR` https://github.com/pingcap/tidb/issues/27977
- [x] #28199

### Information_schema

- [x] #28058
- [x] https://github.com/pingcap/tidb/issues/26681
- [x] conver `primaryRegion` and `regions` to `xxx_constraints` for convenience, refer https://github.com/pingcap/tidb/issues/26681#issuecomment-915738705
- [x] #28798

### Direct Placement Options / General Meta data

(Previously: https://github.com/pingcap/tidb/issues/26581 )

- [x] pingcap/parser#1299
parser: Support the following as _table_options_ in `CREATE|ALTER TABLE` and `CREATE|ALTER DATABASE` contexts. This should also work for partitions, such as `ALTER TABLE t PARTITION p0`.
```
table_option: {
..
 | FOLLOWERS [=] n
 | VOTERS [=] n
 | LEARNERS  [=] n
 | PRIMARY_REGION [=] "str"
 | REGIONS  [=] "str"
 | SCHEDULE [=] "str"
 | CONSTRAINTS  [=] "str"
 | FOLLOWER_CONSTRAINTS  [=] "str"
 | VOTER_CONSTRAINTS  [=] "str"
 | LEARNER_CONSTRAINTS  [=] "str"
 | PLACEMENT POLICY  [=] "str"
}
```

- [x] After the support of new syntax for parser, the old syntax should be removed. https://github.com/pingcap/tidb/issues/27971
- [x] Ensure that table_options are persisted in TiDB DDL, #27799
- [x] Ensure the database_options for placement are persisted in TiDB DDL  https://github.com/pingcap/tidb/issues/27972 @mjonss 
- [x] ALTER DATABASE support https://github.com/pingcap/tidb/issues/27975 @sylzd   
- [x] Ensure the partitionDef options (usually stored as tableOption in ast) for placement are persisted in TiDB DDL https://github.com/pingcap/tidb/issues/27973
- [x] And visible in `SHOW CREATE TABLE` but always enclosed in a version specific comment (`/*T![placement]`). https://github.com/pingcap/tidb/pull/27792
- [x] Ensure that table_options respect "Policy Validation" and "Skipping Policy Validation" sections of proposal. https://github.com/pingcap/tidb/issues/27974

### SHOW PLACEMENT LABELS

- [x] https://github.com/pingcap/tidb/issues/27007
  - @lcwangchao 

### New Placement Rules Engine

(Previously: https://github.com/pingcap/tidb/issues/27011 )

This is a sub-task of Placement Rules in SQL: #18030. Depends on #26580 and #26581. Once metadata has been finished, we need to add a new middle state to all related DDL jobs to sync rules to PD. Previous, we have already implemented similar code, so the work is more of a migration.

This requires the following subtasks:

- [x] #27814, the pre-PR for other subtasks
- [x] alter db/table/partition https://github.com/pingcap/tidb/issues/28833
- [x] flashback/recover table/partition, since metadata is stored in meta, we can revert #20622 and special infoschema treatment for placement rules
* [x] ALTER PLACEMENT POLICY should update placement rules for all tables in scope in PD.
* [x] #27985
* [x] #28159
* [x] #28450 
* [x] Policy Validation
* [x] always get an explicitly configured SCHEDULE option, refer https://github.com/pingcap/tidb/pull/28450#discussion_r717288642
* [x] #28364
* [x] #31000 

### Rule Removal

- [x] https://github.com/pingcap/tidb/issues/18203
- [x] https://github.com/pingcap/tidb/issues/18206
- [x] https://github.com/pingcap/tidb/issues/18208

### Others
- [x] https://github.com/pingcap/tidb/issues/28309
- [x] https://github.com/pingcap/tidb/issues/28766
- [x] https://github.com/pingcap/tidb/issues/28767
- [x] https://github.com/pingcap/tidb/issues/28832
- [x] https://github.com/pingcap/tidb/issues/28834
- [x] https://github.com/pingcap/tidb/issues/28890
- [x] https://github.com/pingcap/tidb/issues/29006
- [x] https://github.com/pingcap/tidb/issues/29088
- [x] https://github.com/pingcap/tidb/issues/29371
- [x] https://github.com/pingcap/tidb/issues/29482
- [x] #29758
- [x] #29821
- [x] #29822
- [x] #29853
- [x] #29986
- [x] #30417
- [x] #30453
- [x] #30741
- [x] #30620
- [x] tikv/pd#4467
- [x] #30454
- [x] #30858, remove isolation level
- [x] #31053
- [x] #31092
- [x] #31127
- [x] #31186 
- [x] #31153
- [x] #31052
- [x] #31271
- [x] #31453
- [x] #31455
- [ ] #31540 
- [x] #31559
- [x] #31702
- [x] #33188

## Remove direct placement options

This is to track the changes required for https://github.com/pingcap/tidb/pull/31423

-  [x] #31622
-  [x] #31668
-  [x] #31425
-  [x] #31426
-  [x] #31427
-  [x] #31485
-  [x] #31486
-  [x] #31424

## Tools Behaviors

- [x] #31760
- [x] #32290
- [x] #32291
- [x] #32828
- [ ] #32292
- [x] #32951

## Future work

We will plan for future delivery:

- [ ] #37251
- [x] #32458
- [ ] #29677
- [ ] Can we use PD placement rules cache? https://github.com/tikv/pd/blob/701b94fb06fb5486958b673f5a344b2e36f0455d/server/config/config.go#L1029-L1030
- [ ] https://github.com/pingcap/tidb/issues/29421.
* [ ] Parser: Add support for PRIMARY_ZONE / ZONES / PRIMARY_RACK / RACKS / PRIMARY_HOST / HOSTS. https://github.com/pingcap/tidb/issues/28697
- [ ] #30415
- [ ] #30452
- [x] #30960
- [ ] probably let user set location_labels?
- [ ] probably recursively delete/unassign placement policies under database, from INCALL users
- [ ] #31001
- [x] #31173
- [ ] #31270
- [x] #31487
- [ ] #35314

**Comment by zz-jason** [comment]
@djshow832 How about changing the `label` keyword to `locality` and changing `COUNT` to `REPLICAS`? With this, a statement that alters the partition policy is:

```sql
ALTER TABLE users ALTER PARTITION west
	ADD PLACEMENT LOCALITY="+zone=west_dc" ROLE="leader" REPLICAS=1;
```

**Comment by zz-jason** [comment]
CockroachDB uses `CONFIGURE ZONE USING` clause, they use `constraints` to specify `label`, `num_replicas` to specify `count`, for example:

```sql
ALTER PARTITION chicago OF INDEX users@*
CONFIGURE ZONE USING
    num_replicas = 3,
    constraints = '{"+region=us-central":1}',
    lease_preferences = '[[+region=us-central]]';
```

See https://www.cockroachlabs.com/docs/stable/topology-geo-partitioned-leaseholders.html for details.

**Comment by djshow832** [comment]
There's a problem for `locality`: when adding tiflash replica through placement rules, it's often specified in the format `"+engine=tiflash"`, it's not locality. @zz-jason

**Comment by zz-jason** [comment]
> There's a problem for `locality`: when adding tiflash replica through placement rules, it's often specified in the format `"+engine=tiflash"`, it's not locality. @zz-jason

We use the following SQL command to add 2 tiflash replicas for table `tpch50`.`lineitem`:

```sql
ALTER TABLE `tpch50`.`lineitem` SET TIFLASH REPLICA 2
```

Let's regard SQL commands as the user interface, `placement rules` in PD as the implementation util.

In this issue, we need to design a SQL command for the user to place a partition of the table to a specific geographic position(Rack, AZ, Region, etc.). Although we are able to use this same SQL command to add TiFlash replicas, it's better not to do it this way. I prefer to use the specific SQL command to do the specific thing.

**Comment by zz-jason** [comment]
I prefer to use `locality`, it's more specific than `constraints`.

**Comment by djshow832** [comment]
So you mean, we do not support adding TiFlash replicas through the `ALTER PARTITION` statement? @zz-jason

**Comment by xhebox** [comment]
The rule cache is introduced to simplify the implementation. But one thing is that, this cache did not drop unnecessary rules when drop table/partitions. Keep a track here.

**Comment by tisonkun** [comment]
@xhebox @morgo would you mind to link the subtasks finished P0 tasks which marked as done here since it is a tracking issue? Also for the remaining P1 features we may follow the tracking issue pattern as in #25899 to build the bidirectional link.

**Comment by xhebox** [comment]
> @xhebox @morgo would you mind to link the subtasks finished P0 tasks which marked as done here since it is a tracking issue? Also for the remaining P1 features we may follow the tracking issue pattern as in #25899 to build the bidirectional link.

I'll do it later. This project will be restarted soon.

**Comment by xhebox** [comment]
@tisonkun Take a look? Almost all PRs and issues are listed. I know PRs are a lot, but as it is not listed in the sub issue, it is listed here.

**Comment by tisonkun** [comment]
@xhebox thanks for your update! That looks good. We don't trace back to all backlogs which gains little but pays a lot.

I'd like to know whether subtasks don't have a dedicated issue will be associated to one before they are started, such as "Support adding placement rules for databases".

**Comment by xhebox** [comment]
@tisonkun It depends, I think. The person who created the issue is scsldb, but the actually project leader is djshow832 and the actual coder is me.

Scsldb created the issue and does update more on this issue. djshow832 created several sub issues. And I did not link the PR to issues somehow. We lack communication and guidance to ensure how the github tracking issue will be maintained. And it is likely the progress checking is done in internal documents.

**Comment by tisonkun** [comment]
@nolouch @djshow832 do you plan to include this feature in v5.2?

**Comment by morgo** [comment]
@tisonkun That is the plan. We will have a better idea on timeframe when we can agree on syntax changes.

Longer story:

The initial proposal handled the 'optimization' use case well, but there were some ambiguous cases that could lead to misconfiguration. We started revisiting the syntax with the intention of better handling compliance requirements (although in the process we've changed a bit more than originally anticipated. But it does improve usability, so even if not entirely intended it is worth it..)

**Comment by nolouch** [comment]
agree, it is worth it :)  and I think this feature needs more time, v5.2 is too tight.

**Comment by xhebox** [comment]
@morgo maybe recover the TODO list back, or make a new tracking issue.

**Comment by tisonkun** [comment]
@morgo @nolouch thanks for your information. Got it.

@morgo I'd like know why you remove all information and add an empty project? There is a discussion about tracking issue vs. project [1](https://tisonkun.github.io/ziggurat/f379bf788b20/#Why-not-GitHub-Project) and I think tracking issue is better to layout subtasks and comment on. What do you think? BTW where removed information goes?

**Comment by morgo** [comment]
> @morgo I'd like know why you remove all information and add an empty project? There is a discussion about tracking issue vs. project [1](https://tisonkun.github.io/ziggurat/f379bf788b20/#Why-not-GitHub-Project) and I think tracking issue is better to layout subtasks and comment on. What do you think? BTW where removed information goes?

I removed the issues because I was attempting to track in a project :-) But I missed the discussion on why not projects, I will convert it to a list here and close the project tomorrow.

**Comment by xxchan** [comment]
Hi @sylzd, I think you should remove the keyword "close" in your PRs so that they do not close this issue...

**Comment by sylzd** [comment]
> Hi @sylzd, I think you should remove the keyword "close" in your PRs so that they do not close this issue...
> 
> @morgo Maybe you can create subissues so that we can link them in PRs (and also see the assigning status)?

Okay~  I thought `close:` will not work, since it has a colon `:` after `close` . 😂
